Add documentation of all prometheus metrics in core Kubernetes (#37522)
kubernetes/website: content/en/docs/reference/instrumentation/metrics.md

> ---
> title: Kubernetes Metrics Across Components
> content_type: instrumentation
> ---

Reviewer: ```suggestion
title: Kubernetes Metrics Reference
content_type: reference
description: >-
  Details of the metric data that Kubernetes components export.
```

## Metrics

> These are the metrics which are exported in Kubernetes components (i.e. kube-apiserver, scheduler, kube-controller-manager, kube-proxy, cloud-controller-manager). 

Reviewer 2: we can add kubelet later if this PR merges before the confirmation & addition of kubelet

> These are the metrics which are exported in Kubernetes components (i.e. kube-apiserver, scheduler, kube-controller-manager, kube-proxy, cloud-controller-manager). 

Author: I've modified the text with @sftim's suggested alternative. It's uploaded now.

> ## Metrics
> 
> These are the metrics which are exported in Kubernetes components (i.e. kube-apiserver, scheduler, kube-controller-manager, kube-proxy, cloud-controller-manager). 
> 
> (auto-generated 2022 Oct 25)

Reviewer: I'd write:
```suggestion
This page details the metrics that different Kubernetes components export. You can query
the metrics endpoint for these components using an HTTP scrape, and fetch the current
metrics data in OpenMetrics format.
```
(not sure if we support OpenMetrics though).

(ideally, also add “scrape” to our glossary)

Author: I want to preserve the generated date so that we can know staleness, do you know where I should stick that bit?

Reviewer: > I want to preserve the generated date so that we can know staleness, do you know where I should stick that bit?

How about using an HTML comment, if that's useful to maintainers.

However, SIG Docs has an existing process where we update reference docs for every major release and not (yet) for minor  releases. So the docs would normally land in a PR for a major release, and then the site generator shows the last update time details at the foot of the page.

> ### List of Kubernetes Metrics

Reviewer: ```suggestion
### List of Kubernetes Metrics
```

> <table class="table">

Reviewer: Consider adding a caption here.

Author: done.

Author: done

<thead>
    <tr>
        <td width="20%">Name</td>
        <td width="10%">Stability Level</td>
        <td width="10%">Type</td>
        <td width="30%">Help</td>
        <td width="20%">Labels</td>

>         <td width="10%">Const Labels</td>

Reviewer: - please use `<th>` elements in this row.
- please give each column a class, to allow CSS styling
- please _don't_ set a width (use SCSS instead)

Reviewer: Future work: also detail which components export each metric.

Author: Yeah can't do this yet, unfortunately, since it's actually not always accessible from the metric definition. But it is on my list of things I want.

Author: Where do I place the scss?

Reviewer: The SCSS goes in `assets/scss`

    </tr>
</thead>
<tbody>

> <tr><td>etcd_version_info</td><td>ALPHA</td><td>Gauge</td><td>Etcd server's binary version</td><td><div>binary_version</div></td><td>None</td></tr>

Reviewer: It'd be handy to set a class on the different `<td>` elements in each row, again to allow custom styling.

Author: What class do you recommend?

Reviewer: An example:
```html
<tr>
  <td class="metric_name">etcd_version_info</td>
  <td class="metric_stability_level" data-stability="alpha">Alpha</td> <!-- use CSS to capitalize if preferred -->
  <td class="metric_time_series_type" data-type="gauge">Gauge</td>
  <td class="metric_description">etcd server's binary version</td>
  <td class="metric_labels_varying"><span class="metric_label">binary_version<span></td>
  <td class="metric_labels_constant">None</td>
</tr>
```

This leaves room for CSS and JavaScript to be added later. It's nice to get it in early and automatically, so that localization teams can minimize their effort when we add features.
